File: ashgf/algorithms/asebo.py
r"""ASEBO: Adaptive Evolution Strategies with Active Subspaces.

ASEBO uses Principal Component Analysis (PCA) on the gradient-history
buffer to identify a low-dimensional *active subspace*, then samples
search directions from a blended covariance matrix that concentrates
sampling probability in the active subspace while maintaining
exploration across the full ambient space.

Mathematical overview
---------------------
Let :math:`G \in \mathbb{R}^{T \times d}` be the buffer of the last
:math:`T` gradient estimates.  After a warm-up of :math:`k` iterations,
we perform PCA on :math:`G` and retain the first
:math:`r = \min\{i \mid \sum_{j=1}^i \lambda_j / \sum \lambda_j \ge \tau\}`
principal components :math:`U_r \in \mathbb{R}^{r \times d}` (rows are
orthonormal).  The covariance used for direction sampling is a convex
blend of the isotropic component and the active-subspace projector:

.. math::

    \Sigma = \sigma \,
    \Bigl(
        \frac{\alpha}{d}\, I_d
        \;+\;
        \frac{1-\alpha}{r}\, U_r^{\mathsf{T}} U_r
    \Bigr),

which satisfies :math:`\operatorname{tr}(\Sigma) = \sigma` for every
choice of :math:`\alpha \in [0,1]`.  Search directions are drawn as

.. math::

    \mathbf{d}_j \sim \mathcal{N}(\mathbf{0}, \Sigma),
    \qquad j = 1, \dots, M,

and the gradient is estimated via antithetic Gaussian smoothing:

.. math::

    \widehat{\nabla f}(\mathbf{x}) =
    \frac{1}{2\sigma M}
    \sum_{j=1}^{M}
    \bigl[
        f(\mathbf{x} + \sigma \mathbf{d}_j)
      - f(\mathbf{x} - \sigma \mathbf{d}_j)
    \bigr] \, \mathbf{d}_j.

The blending parameter :math:`\alpha` is updated adaptively by
comparing the gradient norm projected onto the active subspace
versus its orthogonal complement.

.. note::

    **Design choice – blended covariance vs. probabilistic mixture.**
    The original thesis describes a *probabilistic mixture* where each
    direction is drawn either from the active subspace *or* from the
    isotropic component according to a Bernoulli trial.  This
    implementation uses a *single blended covariance* matrix that
    interpolates the two covariance structures.  Both approaches are
    valid; the blended form is simpler to sample via a single Cholesky
    factorisation and is more amenable to vectorised computation.

References
----------
* [ASEBO] (thesis / paper reference)

"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Callable

import numpy as np

from ashgf.algorithms.base import BaseOptimizer
from ashgf.gradient.estimators import _parallel_eval

# ---------------------------------------------------------------------------
# Optional dependency – scikit-learn is required for PCA-based active
# subspace detection.
# ---------------------------------------------------------------------------
try:
    from sklearn.decomposition import PCA as _PCA

    _HAS_SKLEARN = True
except ImportError:  # pragma: no cover
    _HAS_SKLEARN = False
    _PCA = None  # type: ignore[assignment]

if TYPE_CHECKING:
    # Re-import so the type-checker treats _PCA as the real class.
    from sklearn.decomposition import PCA as _PCA  # noqa: F811

logger = logging.getLogger(__name__)

__all__ = ["ASEBO"]


class ASEBO(BaseOptimizer):
    r"""Adaptive Evolution Strategies with Active Subspaces.

    Uses PCA on the gradient history to identify a low-dimensional
    active subspace, then samples directions from a blended covariance
    matrix.

    Parameters
    ----------
    lr : float
        Fixed learning rate (:math:`\eta`).
    sigma : float
        Base smoothing bandwidth (:math:`\sigma`).
    k : int
        Number of warm-up iterations before performing the first PCA.
    lambd : float
        Regularisation term added to the diagonal of the covariance
        matrix to guarantee positive-definiteness (Tikhonov
        regularisation for the Cholesky factorisation).
    thresh : float
        Explained-variance-ratio threshold for determining the active
        subspace dimension.
    buffer_size : int
        Maximum number of gradient vectors stored in the circular
        history buffer.  Older entries are evicted in FIFO order.
    seed : int
        Random seed for reproducibility.
    eps : float
        Convergence threshold on step-size norm.
    """

    kind = "ASEBO"

    def __init__(
        self,
        lr: float = 1e-4,
        sigma: float = 1e-4,
        k: int = 50,
        lambd: float = 0.1,
        thresh: float = 1e-4,
        buffer_size: int = 200,
        seed: int = 2003,
        eps: float = 1e-8,
    ) -> None:
        super().__init__(seed=seed, eps=eps)

        if not _HAS_SKLEARN:
            raise ImportError(
                "ASEBO requires scikit-learn for PCA-based active-subspace "
                "detection.  Install it with:\n"
                "    pip install scikit-learn"
            )

        if lr <= 0:
            raise ValueError(f"Learning rate must be > 0, got {lr}")
        if sigma <= 0:
            raise ValueError(f"sigma must be > 0, got {sigma}")
        if k < 2:
            raise ValueError(f"k (warm-up) must be at least 2, got {k}")
        if buffer_size < k:
            raise ValueError(f"buffer_size ({buffer_size}) must be >= k ({k})")
        if not 0.0 <= lambd <= 1.0:
            raise ValueError(f"lambd must be in [0, 1], got {lambd}")
        if not 0.0 < thresh < 1.0:
            raise ValueError(f"thresh must be in (0, 1), got {thresh}")

        self.lr = lr
        self.sigma = sigma
        self.k = k
        self.buffer_size = buffer_size
        self.lambd = lambd
        self.thresh = thresh

        # ------------------------------------------------------------------
        # Internal state (initialised in _setup)
        # ------------------------------------------------------------------
        # Circular buffer for gradient history: (buffer_size, dim)
        self._G: np.ndarray | None = None
        self._G_idx: int = 0  # next write position
        self._G_count: int = 0  # number of entries stored (≤ buffer_size)
        self._alpha: float = 1.0  # current blending parameter

    # ------------------------------------------------------------------
    # Template-method hooks
    # ------------------------------------------------------------------

    def _get_step_size(self) -> float:
        return self.lr

    def _setup(self, f: Callable[[np.ndarray], float], dim: int, x: np.ndarray) -> None:
        """Reset internal state before a fresh optimisation run."""
        self._G = np.zeros((self.buffer_size, dim))
        self._G_idx = 0
        self._G_count = 0
        self._alpha = 1.0

    # ------------------------------------------------------------------
    # Gradient estimator
    # ------------------------------------------------------------------

    def grad_estimator(
        self, x: np.ndarray, f: Callable[[np.ndarray], float]
    ) -> np.ndarray:
        """Estimate :math:`\\nabla f(x)` via antithetic Gaussian smoothing
        with directions drawn from the active-subspace-adapted covariance.

        Implements the three critical bug fixes:
          1. divides by ``n_samples`` in the gradient formula,
          2. does **not** normalise directions to unit norm (preserving
             the :math:`\\chi(d)`-distribution required by the thesis),
          3. uses a blended covariance (documented in the class
             docstring).
        """
        dim = len(x)

        # ==============================================================
        # 1. Determine active-subspace dimension and projectors
        # ==============================================================
        if self._G is not None and self._G_count >= self.k:
            # --- PCA on the *valid* portion of the gradient-history buffer ---
            G_valid = self._G[: self._G_count]  # (count, d)
            pca = _PCA()  # type: ignore[misc]  # validated in __init__
            pca.fit(G_valid)

            var_cumsum = np.cumsum(pca.explained_variance_ratio_)
            n_components = int(np.argmax(var_cumsum >= self.thresh) + 1)
            n_components = max(n_components, 10)
            n_components = min(n_components, dim)

            # Active-subspace projector  P_active = U_active^T @ U_active
            U_active = pca.components_[:n_components]  # (r, d)
            P_active = np.dot(U_active.T, U_active)  # (d, d)

            # Number of Monte Carlo directions M
            if self._G_count == self.k:
                # First PCA iteration – use dim directions for stability
                M = dim
            else:
                M = n_components
        else:
            # --- Warm-up phase: isotropic sampling ---
            U_active = np.empty((0, dim))  # empty, for alpha update
            P_active = np.zeros((dim, dim))
            n_components = dim  # placeholder (not used in covariance)
            M = dim

        # ==============================================================
        # 2. Build blended covariance
        #    Σ = σ·[ (α/d)·I_d  +  ((1-α)/r)·P_active ] + λ·I_d
        # ==============================================================
        cov = (self._alpha / dim) * np.eye(dim) + (
            (1.0 - self._alpha) / max(n_components, 1)
        ) * P_active
        cov *= self.sigma

        # Tikhonov regularisation – guarantees Cholesky success
        if self.lambd > 0:
            cov += self.lambd * self.sigma * np.eye(dim)

        # ==============================================================
        # 3. Sample directions from N(0, Σ)
        #    CRITICAL: do NOT normalise to unit length.
        #    The thesis requires E[‖d‖²] = trace(Σ) = σ  (chi-like
        #    distribution), which would be destroyed by normalisation.
        # ==============================================================
        try:
            L = np.linalg.cholesky(cov)  # (d, d) lower-tri
            Z = np.random.standard_normal((dim, M))  # (d, M)
            A = np.dot(L, Z).T  # (M, d)
        except np.linalg.LinAlgError:
            logger.warning(
                "Cholesky decomposition failed – falling back to isotropic directions."
            )
            A = np.random.randn(M, dim)

        # ==============================================================
        # 4. Antithetic gradient estimation (vectorised)
        # ==============================================================
        sigma_A = self.sigma * A  # (M, d) pre-scaled
        points: list[np.ndarray] = []
        for j in range(M):
            d = sigma_A[j]
            points.append(x + d)
            points.append(x - d)

        # Evaluate (parallel if ASHGF_N_JOBS > 1)
        results = _parallel_eval(f, points)

        # ---- Vectorised gradient assembly ----
        # diff[j] = f(x + σ·d_j) - f(x - σ·d_j)
        results_arr = np.asarray(results)  # (2M,)
        diff = results_arr[0::2] - results_arr[1::2]  # (M,)
        grad = np.dot(diff, A) / (2.0 * self.sigma * M)  # (d,) = (M,) @ (M, d)

        # ==============================================================
        # 5. Update circular gradient-history buffer
        # ==============================================================
        # Exponential decay on existing entries
        if self._G_count > 0:
            self._G[: self._G_count] *= 0.99
        self._G[self._G_idx] = grad
        self._G_idx = (self._G_idx + 1) % self.buffer_size
        self._G_count = min(self._G_count + 1, self.buffer_size)

        # ==============================================================
        # 6. Adapt blending parameter α
        #    α ← clip(‖P_ort·g‖ / ‖P_active·g‖, 0, 1)
        #
        #    Optimisation: instead of building the (d×d) matrix P_ort
        #    explicitly, we use the identity
        #        ‖P_ort·g‖² = ‖g‖² - ‖P_active·g‖²
        #    which follows from the Pythagorean theorem since the two
        #    subspaces are orthogonal complements.
        # ==============================================================
        if self._G_count >= self.k and U_active.shape[0] > 0:
            # Projection onto active subspace: g_active = U_active @ g  (r,)
            g_active = np.dot(U_active, grad)  # (r,)
            norm_active_sq = float(np.dot(g_active, g_active))
            norm_total_sq = float(np.dot(grad, grad))
            norm_ort_sq = max(norm_total_sq - norm_active_sq, 0.0)

            norm_active = np.sqrt(norm_active_sq)
            norm_ort = np.sqrt(norm_ort_sq)

            if norm_active > 1e-12:
                self._alpha = norm_ort / norm_active

            self._alpha = float(np.clip(self._alpha, 0.0, 1.0))

        return grad

File: src/algorithms/asebo.rs
//! ASEBO: Adaptive Evolution Strategies with Active Subspaces.
//!
//! Uses PCA (via SVD) on the gradient-history buffer to identify a
//! low-dimensional *active subspace*, then samples search directions
//! from a blended covariance matrix.
//!
//! Mathematical overview (see thesis / Python docstring):
//!
//! Let G ∈ R^{T×d} be the buffer of past gradient estimates.  After
//! a warm-up of `k` iterations, we perform PCA on G and retain the
//! first `r` components such that cumulative explained variance ≥
//! threshold.  The covariance used for direction sampling is:
//!
//!   Σ = σ · [ (α/d)·I_d  +  ((1-α)/r)·U_r^T U_r ] + λ·σ·I_d
//!
//! Directions are drawn from N(0, Σ) and the gradient is estimated
//! via antithetic Gaussian smoothing (directions are NOT normalised).

use ndarray::{Array1, Array2};
use ndarray_linalg::{CholeskyInto, SVDInto, UPLO};

use crate::algorithms::base::Optimizer;
use crate::utils::{parallel_eval, SeededRng};
use rand::Rng;
use rand_distr::StandardNormal;

/// Adaptive Evolution Strategies with Active Subspaces.
pub struct ASEBO {
    /// Fixed learning rate.
    pub lr: f64,
    /// Base smoothing bandwidth.
    pub sigma: f64,
    /// Warm-up iterations before first PCA.
    pub k: usize,
    /// Tikhonov regularisation constant (added to diagonal as λ·σ·I).
    pub lambd: f64,
    /// Explained-variance-ratio threshold for active subspace dimension.
    pub thresh: f64,
    /// Maximum size of the gradient-history circular buffer.
    pub buffer_size: usize,
    /// Number of parallel threads for function evaluation.
    pub n_jobs: usize,
    eps: f64,

    // -- Internal state --
    /// Circular buffer: (buffer_size, dim).
    g_buffer: Option<Array2<f64>>,
    g_idx: usize,
    g_count: usize,
    /// Current blending parameter α ∈ [0, 1].
    current_alpha: f64,
    /// Cached active-subspace basis U_active: (r, dim).
    u_active: Option<Array2<f64>>,
    /// Number of active components r.
    n_components: usize,
    /// Counter for periodic full PCA refit.
    refit_counter: usize,
    /// Number of Monte Carlo directions M.
    m_dir: usize,
}

impl ASEBO {
    pub fn new(
        lr: f64,
        sigma: f64,
        k: usize,
        lambd: f64,
        thresh: f64,
        buffer_size: usize,
        eps: f64,
    ) -> Self {
        assert!(lr > 0.0);
        assert!(sigma > 0.0);
        assert!(k >= 2);
        assert!(buffer_size >= k);
        assert!((0.0..=1.0).contains(&lambd));
        assert!(thresh > 0.0 && thresh < 1.0);
        Self {
            lr,
            sigma,
            k,
            lambd,
            thresh,
            buffer_size,
            n_jobs: 0,
            eps,
            g_buffer: None,
            g_idx: 0,
            g_count: 0,
            current_alpha: 1.0,
            u_active: None,
            n_components: 0,
            refit_counter: 0,
            m_dir: 0,
        }
    }

    /// Perform full PCA via SVD on the gradient buffer.
    /// Returns (U_active, n_components) where U_active is (r, dim).
    fn pca_fit(&self, g_valid: &Array2<f64>, dim: usize) -> (Array2<f64>, usize) {
        let t = g_valid.nrows();
        if t < 2 {
            return (Array2::zeros((0, dim)), 0);
        }

        // Center the data (column-wise mean)
        let mean = g_valid.mean_axis(ndarray::Axis(0)).unwrap();
        let g_centered = g_valid - &mean.insert_axis(ndarray::Axis(0));

        // SVD: G_centered = U Σ V^T  (economy size)
        // V is (dim, dim) with columns = principal directions
        let (_u, sigma, vt_opt) = match g_centered.svd_into(true, true) {
            Ok((u, s, v)) => (u, s, v),
            Err(_) => {
                return (Array2::zeros((0, dim)), 0);
            }
        };

        let vt = match vt_opt {
            Some(v) => v,
            None => return (Array2::zeros((0, dim)), 0),
        };

        // Explained variance from singular values squared
        let var_total: f64 = sigma.iter().map(|&s| s * s).sum();
        if var_total < 1e-30 {
            return (Array2::zeros((0, dim)), 0);
        }

        let var_cumsum: Vec<f64> = sigma
            .iter()
            .map(|&s| s * s / var_total)
            .scan(0.0, |acc, x| {
                *acc += x;
                Some(*acc)
            })
            .collect();

        // Find first index where cumulative variance >= threshold
        let r = var_cumsum
            .iter()
            .position(|&v| v >= self.thresh)
            .map(|p| p + 1)
            .unwrap_or(sigma.len());

        // Clamp: at least 10, at most dim
        let r = r.max(10).min(dim);

        // U_active = first r rows of V^T (= first r columns of V)
        let u_active = vt.slice(ndarray::s![..r, ..]).to_owned();

        (u_active, r)
    }
}

impl Optimizer for ASEBO {
    fn kind(&self) -> &'static str {
        "ASEBO"
    }

    fn eps(&self) -> f64 {
        self.eps
    }

    fn step_size(&self) -> f64 {
        self.lr
    }

    fn setup(&mut self, _f: &(dyn Fn(&Array1<f64>) -> f64 + Sync), dim: usize, _x: &Array1<f64>) {
        self.g_buffer = Some(Array2::zeros((self.buffer_size, dim)));
        self.g_idx = 0;
        self.g_count = 0;
        self.current_alpha = 1.0;
        self.u_active = None;
        self.n_components = 0;
        self.refit_counter = 0;
        self.m_dir = dim;
    }

    fn post_iteration(
        &mut self,
        _iteration: usize,
        _x: &Array1<f64>,
        _grad: &Array1<f64>,
        _f_val: f64,
    ) {
        // ASEBO adapts α inside grad_estimator; no post-iteration logic needed.
    }

    fn grad_estimator(
        &mut self,
        x: &Array1<f64>,
        f: &(dyn Fn(&Array1<f64>) -> f64 + Sync),
        rng: &mut SeededRng,
    ) -> Array1<f64> {
        let dim = x.len();
        let refit_interval = 20;
        let n_jobs = if self.n_jobs > 0 {
            self.n_jobs
        } else {
            rayon::current_num_threads()
        };

        // ==========================================================
        // 1. Determine active-subspace via PCA (after warm-up)
        // ==========================================================
        let (u_active, n_components, m_dir) = if self.g_count >= self.k {
            self.refit_counter += 1;

            let g_valid = self
                .g_buffer
                .as_ref()
                .unwrap()
                .slice(ndarray::s![..self.g_count, ..]);

            if self.refit_counter % refit_interval == 1 || self.u_active.is_none() {
                // Full PCA refit
                let (ua, nc) = self.pca_fit(&g_valid.to_owned(), dim);
                self.u_active = if nc > 0 { Some(ua.clone()) } else { None };
                self.n_components = nc;
            }

            let ua = self.u_active.clone();
            let nc = self.n_components;
            let m = if self.g_count == self.k {
                dim
            } else {
                nc.max(10)
            };

            (ua, nc, m)
        } else {
            // Warm-up: isotropic
            (None, dim, dim)
        };

        self.m_dir = m_dir;

        // ==========================================================
        // 2. Build blended covariance
        //    Σ = σ·[(α/d)·I + ((1-α)/r)·P_active] + λ·σ·I
        // ==========================================================
        // P_active = U_active^T · U_active  (dim, dim)
        let p_active: Array2<f64> = if let Some(ref ua) = u_active {
            if ua.nrows() > 0 {
                ua.t().dot(ua)
            } else {
                Array2::zeros((dim, dim))
            }
        } else {
            Array2::zeros((dim, dim))
        };

        let r_eff = if n_components > 0 { n_components } else { 1 };
        let eye = Array2::<f64>::eye(dim);
        let cov_iso = &eye * (self.current_alpha / dim as f64);
        let cov_active = &p_active * ((1.0 - self.current_alpha) / r_eff as f64);
        let mut cov = (cov_iso + cov_active) * self.sigma;

        // Tikhonov regularisation
        if self.lambd > 0.0 {
            cov = cov + &eye * (self.lambd * self.sigma);
        }

        // ==========================================================
        // 3. Sample directions from N(0, Σ) via Cholesky
        //    A = L @ Z^T  →  rows of A^T = directions (M, dim)
        // ==========================================================
        let a: Array2<f64> = match cov.cholesky_into(UPLO::Lower) {
            Ok(l) => {
                let z = Array2::from_shape_fn((dim, m_dir), |_| rng.rng.sample(StandardNormal));
                let az = l.dot(&z);
                az.t().to_owned() // (M, dim)
            }
            Err(_) => {
                tracing::warn!("ASEBO: Cholesky failed — falling back to isotropic directions.");
                Array2::from_shape_fn((m_dir, dim), |_| rng.rng.sample(StandardNormal))
            }
        };

        // ==========================================================
        // 4. Antithetic gradient estimation
        // ==========================================================
        let sigma_a = &a * self.sigma; // (M, dim) pre-scaled
        let mut points: Vec<Array1<f64>> = Vec::with_capacity(2 * m_dir);
        for j in 0..m_dir {
            let d = sigma_a.row(j).to_owned();
            points.push(x + &d);
            points.push(x - &d);
        }

        let results: Vec<f64> = parallel_eval(f, &points, n_jobs);
        let results_arr = Array1::from_vec(results);

        // diff[j] = f(x + σ·d_j) - f(x - σ·d_j)
        let diff = results_arr.slice(ndarray::s![0..;2]).to_owned()
            - &results_arr.slice(ndarray::s![1..;2]);

        // grad = Σ diff_j · a_j / (2σ·M)
        let grad = diff.dot(&a) / (2.0 * self.sigma * m_dir as f64);

        // ==========================================================
        // 5. Update circular gradient-history buffer (with decay)
        // ==========================================================
        if let Some(ref mut buf) = self.g_buffer {
            // Exponential decay on existing entries
            if self.g_count > 0 {
                for i in 0..self.g_count.min(self.buffer_size) {
                    buf.row_mut(i).mapv_inplace(|v| v * 0.99);
                }
            }
            buf.row_mut(self.g_idx).assign(&grad);
            self.g_idx = (self.g_idx + 1) % self.buffer_size;
            self.g_count = (self.g_count + 1).min(self.buffer_size);
        }

        // ==========================================================
        // 6. Adapt blending parameter α
        //    α = ‖P_ort·g‖ / ‖P_active·g‖  (clipped to [0, 1])
        //    Use: ‖P_ort·g‖² = ‖g‖² - ‖P_active·g‖²
        // ==========================================================
        if self.g_count >= self.k {
            if let Some(ref ua) = u_active {
                if ua.nrows() > 0 {
                    let g_active = ua.dot(&grad); // (r,)
                    let norm_active_sq = g_active.dot(&g_active);
                    let norm_total_sq = grad.dot(&grad);
                    let norm_ort_sq = (norm_total_sq - norm_active_sq).max(0.0);
                    let norm_active = norm_active_sq.sqrt();
                    let norm_ort = norm_ort_sq.sqrt();

                    if norm_active > 1e-12 {
                        self.current_alpha = (norm_ort / norm_active).clamp(0.0, 1.0);
                    }
                }
            }
        }

        grad
    }
}

impl Default for ASEBO {
    fn default() -> Self {
        Self::new(1e-4, 1e-4, 50, 0.1, 1e-4, 200, 1e-8)
    }
}
